File: TracksSample/TracksSample/ViewController.m
//
//  ViewController.m
//  BasicSample
//
//  Created by Gal Orlanczyk on 15/03/2017.
//  Copyright © 2017 Kaltura. All rights reserved.
//

#import "ViewController.h"

@import PlayKit;

@interface ViewController ()

@property (strong, nonatomic) id<Player> player;
@property (strong, nonatomic) NSTimer *playheadTimer;
@property (weak, nonatomic) IBOutlet PlayerView *playerContainer;
@property (weak, nonatomic) IBOutlet UISlider *playheadSlider;

@property (weak, nonatomic) NSArray *audioTracks;
@property (weak, nonatomic) NSArray *textTracks;
@property (weak, nonatomic) NSArray *selectedTracks;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Load the player
    self.player = [PlayKitManager.sharedInstance loadPlayerWithPluginConfig:nil];
    
    // Register events if needed.
    // Event registeration must be after loading the player successfully to make sure events are added, and before prepare to make sure no events are missed (when calling prepare player starts buffering and sends events)
    // Handle audio/ text tracks
    [self handleTracks];
    // Get current bitrate value
    [self handlePlaybackInfo];
    [self handlePlayheadUpdate];
    [self setupTextTrackStyling];
    
    [self setupPlayer];
    
    self.playheadSlider.continuous = NO;
}

- (UIStatusBarStyle)preferredStatusBarStyle {
    return UIStatusBarStyleLightContent;
}

/*********************************/
#pragma mark - Private Methods
/*********************************/

- (void)setupPlayer {
    // Setup the player's view
    self.player.view = self.playerContainer;
    [self.playerContainer sendSubviewToBack:self.player.view];
    
    // Uncomment the type of media needed
//    PKMediaEntry *mediaEntry = [self getMediaWithInternalSubtitles];
    PKMediaEntry *mediaEntry = [self getMediaWithExternalSubtitles];
    
    // Create media config
    MediaConfig *mediaConfig = [[MediaConfig alloc] initWithMediaEntry:mediaEntry startTime:0.0];
    
    // Set if we want the player to auto select the subtitles.
    self.player.settings.trackSelection.textSelectionMode = TrackSelectionModeAuto;
    self.player.settings.trackSelection.textSelectionLanguage = @"en";
    
    [self.player prepare:mediaConfig];
}

- (PKMediaEntry *)getMediaWithInternalSubtitles {
    NSURL *contentURL = [[NSURL alloc] initWithString:@"https://devstreaming-cdn.apple.com/videos/streaming/examples/bipbop_16x9/bipbop_16x9_variant.m3u8"];
    
    // Create media source and initialize a media entry with that source
    NSString *entryId = @"bipbop_16x9";
    PKMediaSource* source = [[PKMediaSource alloc] init:entryId contentUrl:contentURL mimeType:nil drmData:nil mediaFormat:MediaFormatHls];
    NSArray<PKMediaSource*>* sources = [[NSArray alloc] initWithObjects:source, nil];
    // Setup media entry
    PKMediaEntry *mediaEntry = [[PKMediaEntry alloc] init:entryId sources:sources duration:-1];
    
    return mediaEntry;
}

- (PKMediaEntry *)getMediaWithExternalSubtitles {
    NSURL *contentURL = [[NSURL alloc] initWithString:@"https://cdnapisec.kaltura.com/p/2215841/sp/2215841/playManifest/entryId/1_9bwuo813/flavorIds/0_vfdi28n9,1_5j0bgx4v,1_x6tlvn4x,1_zj4vzg46/deliveryProfileId/19201/protocol/https/format/applehttp/a.m3u8"];
    
    // Create media source and initialize a media entry with that source
    NSString *entryId = @"1_9bwuo813";
    PKMediaSource* source = [[PKMediaSource alloc] init:entryId contentUrl:contentURL mimeType:nil drmData:nil mediaFormat:MediaFormatHls];
    NSArray<PKMediaSource*>* sources = [[NSArray alloc] initWithObjects:source, nil];
    // Setup media entry
    PKMediaEntry *mediaEntry = [[PKMediaEntry alloc] init:entryId sources:sources duration:-1];
    
    mediaEntry.externalSubtitles = @[[[PKExternalSubtitle alloc] initWithId:@"Deutsch-de"
                                                                       name:@"Deutsch"
                                                                  isDefault:NO
                                                                 autoSelect:NO
                                                                     forced:NO
                                                                   language:@"de"
                                                            characteristics:nil
                                                               vttURLString:@"http://brenopolanski.com/html5-video-webvtt-example/MIB2-subtitles-pt-BR.vtt"
                                                                   duration:57.0],
                                     [[PKExternalSubtitle alloc] initWithId:@"English-en"
                                                                       name:@"English"
                                                                  isDefault:YES
                                                                 autoSelect:YES
                                                                     forced:NO
                                                                   language:@"en"
                                                            characteristics:nil
                                                               vttURLString:@"http://externaltests.dev.kaltura.com/player/captions_files/eng.vtt"
                                                                   duration:57.0]];
    
    return mediaEntry;
}

// Handle Available Tracks and Present Them
- (void)handleTracks {
    // don't forget to use weak self to prevent retain cycles when needed
    __weak __typeof(self) weakSelf = self;
    
    // add observer to tracksAvailable event
    [self.player addObserver:self events:@[PlayerEvent.tracksAvailable, PlayerEvent.textTrackChanged, PlayerEvent.audioTrackChanged] block:^(PKEvent * _Nonnull event) {
        __strong __typeof(weakSelf) strongSelf = weakSelf;
        
        if ([event isKindOfClass:PlayerEvent.tracksAvailable]) {
            
            // Extract Audio Tracks
            if (event.tracks.audioTracks) {
                strongSelf.audioTracks = event.tracks.audioTracks;
                
                // Set Defualt array for Picker View
                strongSelf.selectedTracks = strongSelf.audioTracks;
            }
            // Extract Text Tracks
            if (event.tracks.textTracks) {
                strongSelf.textTracks = event.tracks.textTracks;
            }
        } else if ([event isKindOfClass:PlayerEvent.textTrackChanged]) {
            NSLog(@"selected text track:: %@", event.selectedTrack.title);
        } else if ([event isKindOfClass:PlayerEvent.audioTrackChanged]) {
            NSLog(@"selected audio track:: %@", event.selectedTrack.title);
        }
    }];
    
    [self printCurrentTrack];
}


// Get Current Audio/ Text Track
- (void)printCurrentTrack {
    NSLog(@"currentAudioTrack:: %@", self.player.currentAudioTrack);
    NSLog(@"currentTextTrack:: %@", self.player.currentTextTrack);
}

// Get Current Bitrate
- (void)handlePlaybackInfo {
    [self.player addObserver:self events:@[PlayerEvent.playbackInfo] block:^(PKEvent * _Nonnull event) {
        if ([event isKindOfClass:PlayerEvent.playbackInfo]) {
            // Get Current Bitrate Value
            if (event.playbackInfo) {
                NSLog(@"%@", event.playbackInfo);
            }
        }
    }];
}

// Select Track
- (void)selectTrack:(Track *)track {
    [self.player selectTrackWithTrackId:track.id];
}

- (void)handlePlayheadUpdate {
    __weak __typeof(self) weakSelf = self;
    
    [self.player addObserver:self events:@[PlayerEvent.playheadUpdate] block:^(PKEvent * _Nonnull event) {
        __strong __typeof(weakSelf) strongSelf = weakSelf;
        [strongSelf playheadUpdate];
    }];
}

- (void)playheadUpdate {
    self.playheadSlider.value = self.player.currentTime / self.player.duration;
}

- (void)setupTextTrackStyling {
    [self.player.settings.textTrackStyling setTextColor:[UIColor redColor]];
    [self.player.settings.textTrackStyling setBackgroundColor:[UIColor whiteColor]];
    [self.player.settings.textTrackStyling setTextSizeWithPercentageOfVideoHeight:10];
    [self.player.settings.textTrackStyling setEdgeStyle:PKTextMarkupCharacterEdgeStyleDropShadow];
    [self.player.settings.textTrackStyling setEdgeColor:[UIColor yellowColor]];
    [self.player.settings.textTrackStyling setFontFamily:@"Helvetica"];
}

/*********************************/
#pragma mark - Picker View
/*********************************/

// The number of columns of data
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

// The number of rows of data
- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return self.selectedTracks.count;
}

// The data to return for the row and component (column) that's being passed in
- (NSString*)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return ((Track *)self.selectedTracks[row]).title;
}

// Catpure the picker view selection
- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    [self selectTrack:((Track *)self.selectedTracks[row])];
}

/*********************************/
#pragma mark - Actions
/*********************************/

- (IBAction)audioTracksTapped:(id)sender {
    self.selectedTracks = self.audioTracks;
    [self.picker reloadAllComponents];
}

- (IBAction)textTracksTapped:(id)sender {
    self.selectedTracks = self.textTracks;
    [self.picker reloadAllComponents];
}

- (IBAction)playTouched:(UIButton *)sender {
    if(!self.player.isPlaying) {
        self.playheadTimer = [NSTimer scheduledTimerWithTimeInterval:0.5f target:self selector:@selector(playheadUpdate) userInfo:nil repeats:YES];
        [self.player play];
    }
}

- (IBAction)pauseTouched:(UIButton *)sender {
    if(self.player.isPlaying) {
        [self.playheadTimer invalidate];
        self.playheadTimer = nil;
        [self.player pause];
    }
}

- (IBAction)playheadValueChanged:(UISlider *)sender {
    NSLog(@"playhead value: %f", sender.value);
    self.player.currentTime = self.player.duration * sender.value;
}

@end

File: TracksSample/TracksSample-swift/ViewController.swift
//
//  ViewController.swift
//  TracksSample-swift
//
//  Created by Vadik on 07/11/2017.
//  Copyright © 2017 kaltura. All rights reserved.
//

import UIKit
import PlayKit

class ViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSource {

    @IBOutlet weak var playerContainer: PlayerView!
    @IBOutlet weak var playheadSlider: UISlider!
    @IBOutlet weak var picker: UIPickerView!
    
    var player: Player?
    var audioTracks: [Track] = []
    var textTracks: [Track] = []
    var selectedTracks: [Track] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        // 1. Load the player
        player = PlayKitManager.shared.loadPlayer(pluginConfig: nil)
        // 2. Register events if needed.
        // Event registeration must be after loading the player successfully to make sure events are added,
        // and before prepare to make sure no events are missed (when calling prepare player starts buffering and sending events)
        handleTracks()
        handlePlaybackInfo()
        handlePlayheadUpdate()
        
        // 3. Prepare the player (can be called at a later stage, preparing starts buffering the video)
        preparePlayer()
    }
    
    override var preferredStatusBarStyle : UIStatusBarStyle {
        return UIStatusBarStyle.lightContent
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        switch segue.identifier {
        case "TextStylingSegue":
            guard let navigationVC = segue.destination as? UINavigationController else { return }
            guard let textStylingVC = navigationVC.topViewController as? TextStylingViewController else { return }
            textStylingVC.player = player
        default:
            return
        }
    }

    func preparePlayer() {
        // Setup the player's view
        player?.view = self.playerContainer
        playerContainer.sendSubviewToBack(self.player!.view!)
       
        // Uncomment the type of media needed
//        let mediaEntry = getMediaWithInternalSubtitles()
        let mediaEntry = getMediaWithExternalSubtitles()
//        let mediaEntry = getDRMMediaWithExternalSubtitles()
        
        // Create media config
        let mediaConfig = MediaConfig(mediaEntry: mediaEntry)
        
        // Set if we want the player to auto select the subtitles.
        player?.settings.trackSelection.textSelectionMode = .auto
        player?.settings.trackSelection.textSelectionLanguage = "en"
        
        // Prepare the player
        player?.prepare(mediaConfig)
    }
    
    func getMediaWithInternalSubtitles() -> PKMediaEntry {
        let contentURL = "https://devstreaming-cdn.apple.com/videos/streaming/examples/bipbop_16x9/bipbop_16x9_variant.m3u8"
        
        // Create media source and initialize a media entry with that source
        let entryId = "bipbop_16x9"
        let source = PKMediaSource(entryId, contentUrl: URL(string: contentURL), drmData: nil, mediaFormat: .hls)
        
        // Setup media entry
        let mediaEntry = PKMediaEntry(entryId, sources: [source], duration: -1)
        // If we set external subtitles, they will be ignored.
        
        return mediaEntry
    }
    
    func getMediaWithExternalSubtitles() -> PKMediaEntry {
        let contentURL = "https://cdnapisec.kaltura.com/p/2215841/sp/2215841/playManifest/entryId/1_9bwuo813/flavorIds/0_vfdi28n9,1_5j0bgx4v,1_x6tlvn4x,1_zj4vzg46/deliveryProfileId/19201/protocol/https/format/applehttp/a.m3u8"
        
        // Create media source and initialize a media entry with that source
        let entryId = "1_9bwuo813"
        let source = PKMediaSource(entryId, contentUrl: URL(string: contentURL), drmData: nil, mediaFormat: .hls)
        
        // Setup media entry
        let mediaEntry = PKMediaEntry(entryId, sources: [source], duration: -1)
        mediaEntry.externalSubtitles = [PKExternalSubtitle(id: "Deutsch-de",
                                                           name: "Deutsch",
                                                           language: "de",
                                                           vttURLString: "http://brenopolanski.com/html5-video-webvtt-example/MIB2-subtitles-pt-BR.vtt",
                                                           duration: 570.0),
                                        PKExternalSubtitle(id: "English-en",
                                                           name: "English",
                                                           language: "en",
                                                           vttURLString: "http://externaltests.dev.kaltura.com/player/captions_files/eng.vtt",
                                                           duration: 570.0)]
        
        return mediaEntry
    }
    
    func getDRMMediaWithExternalSubtitles() -> PKMediaEntry {
        let contentURL = "https://cdnapisec.kaltura.com/p/2222401/sp/2222401/playManifest/entryId/1_i18rihuv/flavorIds/1_nwoofqvr,1_3z75wwxi,1_exjt5le8,1_uvb3fyqs/deliveryProfileId/8642/protocol/https/format/applehttp/a.m3u8"
        
        // Create media source and initialize a media entry with that source
        let entryId = "0_pl5lbfo0"
        let drmParams = FairPlayDRMParams(licenseUri: "https://udrmv3.kaltura.com/fps/license?custom_data=eyJjYV9zeXN0ZW0iOiJPVlAiLCJ1c2VyX3Rva2VuIjoiZGpKOE1qSXlNalF3TVh3amRkODBoRzFnNEVFU1ZLdG9KVmF6VlJRWFEwZ05faDRLRGxvOVc5NklEOUw1c0hLdGkwZ1J3UXM5dW5zbTlLTmNOT1o2QWtDR0hFRTdlMmZuNHFOdGlPYjM1M2k4QTZaQWlnbktxS0hsNHc9PSIsImFjY291bnRfaWQiOiIyMjIyNDAxIiwiY29udGVudF9pZCI6IjFfaTE4cmlodXYiLCJmaWxlcyI6IjFfbndvb2ZxdnIsMV8zejc1d3d4aSwxX2V4anQ1bGU4LDFfdXZiM2Z5cXMifQ%3D%3D&signature=KbGUO7ucxk7xDIdIWS4IKtAdSZU%3D",
                                          base64EncodedCertificate: "MIIFETCCA/mgAwIBAgIISWLo8KcYfPMwDQYJKoZIhvcNAQEFBQAwfzELMAkGA1UEBhMCVVMxEzARBgNVBAoMCkFwcGxlIEluYy4xJjAkBgNVBAsMHUFwcGxlIENlcnRpZmljYXRpb24gQXV0aG9yaXR5MTMwMQYDVQQDDCpBcHBsZSBLZXkgU2VydmljZXMgQ2VydGlmaWNhdGlvbiBBdXRob3JpdHkwHhcNMTYwMjAxMTY0NTQ0WhcNMTgwMjAxMTY0NTQ0WjCBijELMAkGA1UEBhMCVVMxKDAmBgNVBAoMH1ZJQUNPTSAxOCBNRURJQSBQUklWQVRFIExJTUlURUQxEzARBgNVBAsMClE5QU5HR0w4TTYxPDA6BgNVBAMMM0ZhaXJQbGF5IFN0cmVhbWluZzogVklBQ09NIDE4IE1FRElBIFBSSVZBVEUgTElNSVRFRDCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA2YmfdPWM86+te7Bbt4Ic6FexXwMeL+8AmExIj8jAaNxhKbfVFnUnuXzHOajGC7XDbXxsFbPqnErqjw0BqUoZhs+WVMy+0X4AGqHk7uRpZ4RLYganel+fqitL9rz9w3p41x8JfLV+lAej+BEN7zNeqQ2IsC4BxkViu1gA6K22uGsCAwEAAaOCAgcwggIDMB0GA1UdDgQWBBQK+Gmarl2PO3jtLP6A6TZeihOL3DAMBgNVHRMBAf8EAjAAMB8GA1UdIwQYMBaAFGPkR1TLhXFZRiyDrMxEMWRnAyy+MIHiBgNVHSAEgdowgdcwgdQGCSqGSIb3Y2QFATCBxjCBwwYIKwYBBQUHAgIwgbYMgbNSZWxpYW5jZSBvbiB0aGlzIGNlcnRpZmljYXRlIGJ5IGFueSBwYXJ0eSBhc3N1bWVzIGFjY2VwdGFuY2Ugb2YgdGhlIHRoZW4gYXBwbGljYWJsZSBzdGFuZGFyZCB0ZXJtcyBhbmQgY29uZGl0aW9ucyBvZiB1c2UsIGNlcnRpZmljYXRlIHBvbGljeSBhbmQgY2VydGlmaWNhdGlvbiBwcmFjdGljZSBzdGF0ZW1lbnRzLjA1BgNVHR8ELjAsMCqgKKAmhiRodHRwOi8vY3JsLmFwcGxlLmNvbS9rZXlzZXJ2aWNlcy5jcmwwDgYDVR0PAQH/BAQDAgUgMEgGCyqGSIb3Y2QGDQEDAQH/BDYBZ2diOGN5bXpsb21vdXFqb3p0aHg5aXB6dDJ0bThrcGdqOGNwZGlsbGVhMWI1aG9saWlyaW8wPQYLKoZIhvdjZAYNAQQBAf8EKwF5aHZlYXgzaDB2Nno5dXBqcjRsNWVyNm9hMXBtam9zYXF6ZXdnZXFkaTUwDQYJKoZIhvcNAQEFBQADggEBAIaTVzuOpZhHHUMGd47XeIo08E+Wb5jgE2HPsd8P/aHwVcR+9627QkuAnebftasV/h3FElahzBXRbK52qIZ/UU9nRLCqqKwX33eS2TiaAzOoMAL9cTUmEa2SMSzzAehb7lYPC73Y4VQFttbNidHZHawGp/844ipBS7Iumas8kT8G6ZmIBIevWiggd+D5gLdqXpOFI2XsoAipuxW6NKnnlKnuX6aNReqzKO0DmQPDHO2d7pbd3wAz5zJmxDLpRQfn7iJKupoYGqBs2r45OFyM14HUWaC0+VSh2PaZKwnSS8XXo4zcT/MfEcmP0tL9NaDfsvIWnScMxHUUTNNsZIp3QXA=")
        
        let source = PKMediaSource(entryId, contentUrl: URL(string: contentURL), drmData: [drmParams], mediaFormat: .hls)
        
        // Setup media entry
        let mediaEntry = PKMediaEntry(entryId, sources: [source], duration: -1)
        mediaEntry.externalSubtitles = [PKExternalSubtitle(id: "Deutsch-de",
                                                           name: "Deutsch",
                                                           language: "de",
                                                           vttURLString: "http://brenopolanski.com/html5-video-webvtt-example/MIB2-subtitles-pt-BR.vtt",
                                                           duration: 888.11),
                                        PKExternalSubtitle(id: "English-en",
                                                           name: "English",
                                                           language: "en",
                                                           vttURLString: "http://externaltests.dev.kaltura.com/player/captions_files/eng.vtt",
                                                           duration: 888.11)]
        
        return mediaEntry
    }
    
    func handleTracks() {
        player?.addObserver(self, events: [PlayerEvent.tracksAvailable, PlayerEvent.textTrackChanged, PlayerEvent.audioTrackChanged], block: { [weak self] (event: PKEvent) in
            if type(of: event) == PlayerEvent.tracksAvailable {
                guard let this = self else { return }
                
                if let tracks = event.tracks?.audioTracks {
                    this.audioTracks = tracks
                    this.selectedTracks = this.audioTracks
                }
                
                if let tracks = event.tracks?.textTracks {
                    this.textTracks = tracks
                }
                
                this.picker.reloadAllComponents()
            } else if type(of: event) == PlayerEvent.textTrackChanged {
                print("selected text track: \(event.selectedTrack?.title ?? "")")
            } else if type(of: event) == PlayerEvent.audioTrackChanged {
                print("selected audio track: \(event.selectedTrack?.title ?? "")")
            }
        })
    }
    
    func handlePlaybackInfo() {
        player?.addObserver(self, event: PlayerEvent.playbackInfo) { event in
            if type(of: event) == PlayerEvent.playbackInfo {
                if let _ = event.playbackInfo {
                    print("\(event.playbackInfo!)")
                }
            }
        }
    }
    
    func handlePlayheadUpdate() {
        player?.addObserver(self, event: PlayerEvent.playheadUpdate, block: { [weak self] (event) in
            guard let self = self else { return }
            self.playheadUpdate()
        })
    }
    
    func selectTrack(_ track: Track) {
        player?.selectTrack(trackId: track.id)
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return selectedTracks.count
    }

    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return selectedTracks[row].title
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        if !selectedTracks.isEmpty {
            selectTrack(selectedTracks[row])
        }
    }
    
    @objc func playheadUpdate() {
        if let _ = player {
            playheadSlider.value = Float(player!.currentTime / player!.duration)
        }
    }
    
    // MARK: - IBAction
    
    @IBAction func didSeek(_ sender: UISlider) {
        print("playhead value: \(sender.value)")
        if let _ = player {
            player!.currentTime = player!.duration * TimeInterval(sender.value)
        }
    }
    
    @IBAction func didTapPause(_ sender: UIButton) {
        if player?.isPlaying == true {
            player?.pause()
        }
    }
    
    @IBAction func didTapPlay(_ sender: UIButton) {
        if player?.isPlaying == false {
            player?.play()
        }
    }
    
    @IBAction func didTapAudio(_ sender: UIButton) {
        selectedTracks = audioTracks
        picker.reloadAllComponents()
    }
    
    @IBAction func didTapText(_ sender: UIButton) {
        selectedTracks = textTracks
        picker.reloadAllComponents()
    }
}
